File: src/09.tree/38.297.ts
import { TreeNode } from "./01.100";

/*
 * Encodes a tree to a single string.
 */
export function serialize(root: TreeNode | null): string {
    const ans: string[] = [];
    const dfs = (node: TreeNode | null) => {
        if (!node) {
            ans.push("null");
            return;
        }
        ans.push(node.val + "");
        dfs(node.left);
        dfs(node.right);
    };

    dfs(root);

    return ans.join(",");
}

/*
 * Decodes your encoded data to tree.
 */
export function deserialize(data: string): TreeNode | null {
    const queue = data.split(",");
    const dfs = (arr: string[]): TreeNode | null => {
        const v = arr.shift();
        if (v === "null") {
            return null;
        }

        const node = new TreeNode(Number(v));
        node.left = dfs(arr);
        node.right = dfs(arr);
        return node;
    };

    return dfs(queue);
}

File: src/09.tree/38.297.go
package tree

import (
	"strconv"
	"strings"
)

// Serializes a tree to a single string.
func Serialize(root *TreeNode) string {
	var dfs func(node *TreeNode)
	ans := []string{}

	dfs = func(node *TreeNode) {
		if node == nil {
			ans = append(ans, "null")
			return
		}

		ans = append(ans, strconv.Itoa(node.Val))
		dfs(node.Left)
		dfs(node.Right)
	}

	dfs(root)

	return strings.Join(ans, ",")
}

// Deserializes your encoded data to tree.
func Deserialize(data string) *TreeNode {
	var dfs func() *TreeNode
	queue := strings.Split(data, ",")

	dfs = func() *TreeNode {
		v := queue[0]
		queue = queue[1:]

		if v == "null" {
			return nil
		}

		c, _ := strconv.Atoi(v)
		node := &TreeNode{Val: c}
		node.Left = dfs()
		node.Right = dfs()
		return node
	}

	return dfs()
}
